.github/workflows/verify.yml

```
@@ -16,3 +16,19 @@ jobs:
       run: npm ci
     - name: Build font
       run: npm run build
+  testpage:
+    name: Test page
+    needs: build
+    runs-on: ubuntu-latest
+    steps:
+    - name: Checkout
+      uses: actions/checkout@v2
+    - name: Use Node.js 12.x
+      uses: actions/setup-node@v2
+      with:
+        node-version: 12.x
+    - name: Install dependencies
+      run: npm ci
+    - name: Build font
+      run: npm run build:testpage
```

What about build the test page in the build step? Adding another `testpage` script maybe?

```suggestion
    - name: Build test page
      run: node scripts/buildTestpage.js
```

---

In addition, to recover the state of the previous build, you only need to upload the `font/` folder as an artifact and download it in the "Test page" step.

---

> What about build the test page in the "Build" step? Adding another testpage script maybe?

We could do that, I'm not against it. The reason I choose to do it like this is because I don't want to dig into the CI logs to figure out whether `npm run build` failed or `npm run build:testpage`.


> In addition, to recover the state of the previous build, you only need to upload the `font/` folder as an artifact and download it in the "Test page" step.

I guess we also need `node_modules/`, but maybe that can be cached? Building (including the test page) currently takes 7 seconds and installing dependencies takes 20 seconds, so if I'd say we first figure out how to get that close to 0 - if we keep 2 jobs to begin with.

---

OK, I think that we can try uploading `font/` and `node_modules/` as artifacts. You guide yourself on how to do it @ericcornelissen?

---

Do you think uploading the `node_modules/` as artefacts is the way to go, that seems odd to me and a bit of an abuse of the artefacts functionality :sweat_smile: Why not cache it?

---

But caching it means that is keeped between workflows, not?

---

[actions/cache does not recommend caching `node_modules/`](https://github.com/actions/cache/blob/main/examples.md#node---npm).

---

> But caching it means that is keeped between workflows, not?

As you can see in the CI logs for f54747a33f13fa73cad0ed24c53b28aabcb4668a, caching between jobs works just fine. The reason I would prefer this over artefacts is that artefacts are probably more expensive to upload than whatever [actions/cache](https://github.com/actions/cache) does.


> actions/cache does not recommend caching node_modules/: "it can break across Node versions and won't work with npm ci".

Hmm, I should have remember that :sweat_smile: Actually, interesting that [the documentation](https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows) provides an example with NPM! 

But I suppose we can work around that, e.g. by using `npm install` instead of `npm ci`. Also, we don't work across Node versions to begin with.

---

> Do you think uploading the node_modules/ as artefacts is the way to go, that seems odd to me and a bit of an abuse of the artefacts functionality

Why do you think that? In essence, is the same process of cache (upload / download files), but it's not keeped between workflows, so from the point of view of disk usage, caching is even more abussive.

---

> In essence, is the same process of cache (upload / download files)

I agree with that, from that perspective it doesn't seem to make a difference. Given the remainder of my comment below, I would conclude the following: cache vs artefacts should not make much of a difference, since we don't want to download the `node_modules` I would say it makes sense to not expose them as artefacts.

> but it's not keeped between workflows, so from the point of view of disk usage, caching is even more abussive.

I'm not sure if I follow. Artefacts are definitely uploaded somewhere else as they are, among other things intended, for download by humans. And I don't see any indication that they're kept between job runs.

On the other side, f54747a33f13fa73cad0ed24c53b28aabcb4668a shows that the run of the testpage job uses the cache of the build job. Sure, the cache needs to be reloaded, but I don't see how that makes disk usage _more_ abusive :thinking:

---

Also: https://docs.github.com/en/free-pro-team@latest/actions/guides/caching-dependencies-to-speed-up-workflows#comparing-artifacts-and-dependency-caching